crates/uv-resolver/src/universal_marker.rs

```diff
@@ -106,26 +252,359 @@ impl UniversalMarker {
+        let mut marker = self.marker;
+        marker.implies(other.marker);
+        ConflictMarker { marker }
+    }
+
+    /// Returns true if this conflict marker will always evaluate to `true`.
+    pub fn is_true(self) -> bool {
+        self.marker.is_true()
+    }
+
+    /// Returns true if this conflict marker will always evaluate to `false`.
+    pub fn is_false(self) -> bool {
+        self.marker.is_false()
+    }
+
+    /// Returns true if this conflict marker is satisfied by the given
+    /// list of activated extras and groups.
+    pub(crate) fn evaluate(
+        self,
+        extras: &[(PackageName, ExtraName)],
+        groups: &[(PackageName, GroupName)],
+    ) -> bool {
+        static DUMMY: std::sync::LazyLock<MarkerEnvironment> = std::sync::LazyLock::new(|| {
+            MarkerEnvironment::try_from(MarkerEnvironmentBuilder {
+                implementation_name: "",
+                implementation_version: "3.7",
+                os_name: "linux",
+                platform_machine: "",
+                platform_python_implementation: "",
+                platform_release: "",
+                platform_system: "",
+                platform_version: "",
+                python_full_version: "3.7",
+                python_version: "3.7",
+                sys_platform: "linux",
+            })
+            .unwrap()
+        });
+        let extras = extras
+            .iter()
+            .map(|(package, extra)| encode_package_extra(package, extra));
+        let groups = groups
+            .iter()
+            .map(|(package, group)| encode_package_group(package, group));
+        self.marker
+            .evaluate(&DUMMY, &extras.chain(groups).collect::<Vec<ExtraName>>())
+    }
+}
+
+impl std::fmt::Debug for ConflictMarker {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        // This is a little more succinct than the default.
+        write!(f, "ConflictMarker({:?})", self.marker)
+    }
+}
+
+/// Encodes the given package name and its corresponding extra into a valid
+/// `extra` value in a PEP 508 marker.
+fn encode_package_extra(package: &PackageName, extra: &ExtraName) -> ExtraName {
+    // This is OK because `PackageName` and `ExtraName` have the same
+    // validation rules, and we combine them in a way that always results in a
+    // valid name.
+    //
+    // Note also that we encode the length of the package name (in bytes) into
+    // the encoded extra name as well. This ensures we can parse out both the
+    // package and extra name if necessary. If we didn't do this, then some
+    // cases could be ambiguous since our field delimiter (`-`) is also a valid
+    // character in `package` or `extra` values. But if we know the length of
+    // the package name, we can always parse each field unambiguously.
+    let package_len = package.as_str().len();
+    ExtraName::new(format!("extra-{package_len}-{package}-{extra}")).unwrap()
```

Comment:
Ohh ok. I see now. It's sort of unfortunate that this representation is "user-facing" in the lockfile.

Reply:
I don't know that I have a better answer. We could do something that's legal but not normalized...? Like use dashes for one piece?

Reply:
Can you elaborate?

Reply:
And yes, I 100% agree that this is unfortunate. I just don't know of a better way to go about it without inventing our own marker system.

Reply:
Are we bound to PEP 508, or could we do something like adding a sigil that only allowed here and in the lockfile but gets rejected everywhere else, or even as a specific new marker type? I see the problem with the MarkerTree being nested and such, but this looks like a potential collision and it's user facing

Reply:
> but this looks like a potential collision and it's user facing

I find this very difficult to reason about. On the one hand, it is certainly true that someone could use `extra-3-pkg-foo` as an extra name, and thus potentially collide with a separately defined extra named `foo` for the package `pkg`. (I think such cases would be pathological given the baroqueness of `extra-3-pkg-foo`. You not only need an extra with that name but it _also_ has to be declared as a conflicting extra.) On the other hand, even if a collision does occur, it's actually not totally clear to me that it would cause any problems. That's because we tend to strip the "normal" `extra == 'foo'` from markers at various points.

> Are we bound to PEP 508, or could we do something like adding a sigil that only allowed here and in the lockfile but gets rejected everywhere else, or even as a specific new marker type?

A new marker type is possible (I think I suggested a new marker type or even something that isn't a marker somewhere), but I believe it's costly in terms of implementation effort. And even if we do have a new marker type, there is enough inter-operation going on that we'd need to be able to convert between the marker types. I think it's doable, but yeah, a fair bit of work.

Some extra sigil that isn't typically allowed in extra names is also possible... I think it's probably less work than adding a new marker type. And the [implementation kinda-sorta already makes room for it](https://github.com/astral-sh/uv/blob/e9d545e4a6257337b399cfa2f87d88018f42c174/crates/uv-pep508/src/marker/tree.rs#L421-L428). There's definitely some refactoring cost involved here because we wouldn't be able to use `ExtraName` or `GroupName`. So that means new APIs for evaluating markers with something looser than `&[ExtraName]`. And new parsing APIs to "allow" the sigil in certain cases but continue to reject it in others. Overall possible, but I'd say it's at least a couple days of work.

Since I think that, in order for a collision to occur, it has to be pathological, I'm not sure it's worth doing. One thing I can do as a mitigation is add a test that specifically provokes a collision and ensure behavior is reasonable.

Reply:
> Are we bound to PEP 508

I think we are bound to PEP 508 insomuch as we want to be. While these conflict markers are _technically_ valid PEP 508 markers, I think they are definitely a step toward PEP 508 incompatibility since they require an encoding scheme to use properly.

Reply:
I've added a test that instigates a collision and things seem to be okay:

https://github.com/astral-sh/uv/blob/cc1d8e62ef3d7d84bc687bc5d4ca219d16b2b5ed/crates/uv/tests/it/lock_conflict.rs#L4795-L5015

I think it's because we don't tend to have both `extra == "foo"` and `extra == "pkg-3-foo-extra"` in the same marker at the same time. (And I say "tend to" because I find it _very_ hard to reason about this.)